File: ios/Wrappers/EnrichedMessageQueryParamsWrapper.swift
import Foundation
import XMTP

struct EnrichedMessageQueryParamsWrapper {
	let limit: Int?
	let beforeNs: Int64?
	let afterNs: Int64?
	let direction: String?
	let excludeSenderInboxIds: [String]?
	let deliveryStatus: String?
	let insertedAfterNs: Int64?
	let insertedBeforeNs: Int64?
	let sortBy: String?

	static func fromJson(_ paramsJson: String) -> EnrichedMessageQueryParamsWrapper {
		guard !paramsJson.isEmpty else {
			return EnrichedMessageQueryParamsWrapper(
				limit: nil,
				beforeNs: nil,
				afterNs: nil,
				direction: nil,
				excludeSenderInboxIds: nil,
				deliveryStatus: nil,
				insertedAfterNs: nil,
				insertedBeforeNs: nil,
				sortBy: nil
			)
		}

		let data = paramsJson.data(using: .utf8) ?? Data()
		let jsonOptions =
			(try? JSONSerialization.jsonObject(with: data, options: []))
			as? [String: Any] ?? [:]

		let limit = jsonOptions["limit"] as? Int
		let beforeNs = jsonOptions["beforeNs"] as? Int64
		let afterNs = jsonOptions["afterNs"] as? Int64
		let direction = jsonOptions["direction"] as? String
		let excludeSenderInboxIds = jsonOptions["excludeSenderInboxIds"] as? [String]
		let deliveryStatus = jsonOptions["deliveryStatus"] as? String
		let insertedAfterNs = jsonOptions["insertedAfterNs"] as? Int64
		let insertedBeforeNs = jsonOptions["insertedBeforeNs"] as? Int64
		let sortBy = jsonOptions["sortBy"] as? String

		return EnrichedMessageQueryParamsWrapper(
			limit: limit,
			beforeNs: beforeNs,
			afterNs: afterNs,
			direction: direction,
			excludeSenderInboxIds: excludeSenderInboxIds,
			deliveryStatus: deliveryStatus,
			insertedAfterNs: insertedAfterNs,
			insertedBeforeNs: insertedBeforeNs,
			sortBy: sortBy
		)
	}
}

File: android/src/main/java/expo/modules/xmtpreactnativesdk/wrappers/EnrichedMessageQueryParamsWrapper.kt
package expo.modules.xmtpreactnativesdk.wrappers

import com.google.gson.JsonParser

class EnrichedMessageQueryParamsWrapper(
    val limit: Int?,
    val beforeNs: Long?,
    val afterNs: Long?,
    val direction: String?,
    val excludeSenderInboxIds: List<String>?,
    val deliveryStatus: String?,
    val insertedAfterNs: Long?,
    val insertedBeforeNs: Long?,
    val sortBy: String?,
) {
    companion object {
        fun fromJson(paramsJson: String): EnrichedMessageQueryParamsWrapper {
            if (paramsJson.isEmpty()) {
                return EnrichedMessageQueryParamsWrapper(
                    limit = null,
                    beforeNs = null,
                    afterNs = null,
                    direction = null,
                    excludeSenderInboxIds = null,
                    deliveryStatus = null,
                    insertedAfterNs = null,
                    insertedBeforeNs = null,
                    sortBy = null,
                )
            }

            val jsonOptions = JsonParser.parseString(paramsJson).asJsonObject

            val limit =
                if (jsonOptions.has("limit")) {
                    jsonOptions.get("limit").asInt
                } else {
                    null
                }

            val beforeNs =
                if (jsonOptions.has("beforeNs")) {
                    jsonOptions.get("beforeNs").asLong
                } else {
                    null
                }

            val afterNs =
                if (jsonOptions.has("afterNs")) {
                    jsonOptions.get("afterNs").asLong
                } else {
                    null
                }

            val direction =
                if (jsonOptions.has("direction")) {
                    jsonOptions.get("direction").asString
                } else {
                    null
                }

            val excludeSenderInboxIds =
                if (jsonOptions.has("excludeSenderInboxIds")) {
                    val idsArray = jsonOptions.getAsJsonArray("excludeSenderInboxIds")
                    idsArray.map { it.asString }
                } else {
                    null
                }

            val deliveryStatus =
                if (jsonOptions.has("deliveryStatus")) {
                    jsonOptions.get("deliveryStatus").asString
                } else {
                    null
                }

            val insertedAfterNs =
                if (jsonOptions.has("insertedAfterNs")) {
                    jsonOptions.get("insertedAfterNs").asLong
                } else {
                    null
                }

            val insertedBeforeNs =
                if (jsonOptions.has("insertedBeforeNs")) {
                    jsonOptions.get("insertedBeforeNs").asLong
                } else {
                    null
                }

            val sortBy =
                if (jsonOptions.has("sortBy")) {
                    jsonOptions.get("sortBy").asString
                } else {
                    null
                }

            return EnrichedMessageQueryParamsWrapper(
                limit = limit,
                beforeNs = beforeNs,
                afterNs = afterNs,
                direction = direction,
                excludeSenderInboxIds = excludeSenderInboxIds,
                deliveryStatus = deliveryStatus,
                insertedAfterNs = insertedAfterNs,
                insertedBeforeNs = insertedBeforeNs,
                sortBy = sortBy,
            )
        }
    }
}
